product.py

```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
         self.mount_state()
         result = True
         try:
+            # Don't propose to upgrade a BIOS installation with a UEFI installer and conversely
+            tool = PartitionTool(self.primary_disk)
+            existing_boot_mode = constants.TARGET_BOOT_MODE_LEGACY
+            boot_partnum = tool.partitionNumber(self.boot_device)
+            boot_part = tool.getPartition(boot_partnum)
+            if 'id' in boot_part and boot_part['id'] == GPTPartitionTool.ID_EFI_BOOT:
+                existing_boot_mode = constants.TARGET_BOOT_MODE_UEFI
+
+            if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI and not constants.UEFI_INSTALLER) or \
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
+                    logger.log("Cannot upgrade %s, installer mode (%s) is mismatched with existing boot mode (%s)" %
+                                (self.primary_disk, "UEFI" if constants.UEFI_INSTALLER else "legacy", existing_boot_mode))
+                    raise Exception("Boot mode mismatch")
```

Comment:
`return False` would me more direct, this descriptive string is not used

Reply:
Done